### PR TITLE
Transformers agents 2.0

### PR DESCRIPTION
**Pull request by aymeric-roucher**

### CODE REVIEW

**Comment by aymeric-roucher** [comment]
@LysandreJik here is the blog post PR

**Comment by aymeric-roucher** [comment]
Images are added! ✅  Everything is ready for review.

File: agents.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+Now it’s straightforward to create an agent that leverages this tool!
+
+By the way, the tool could have been loaded from the hub. The documentation shows how to use it, basically: 
+
+```python
+from transformers.agents import HfEngine, ReactJsonAgent
+
+llm_engine = HfEngine("meta-llama/Meta-Llama-3-70B-Instruct")
+
+agent = ReactJsonAgent(tools=[RetrieverTool(vectordb, all_sources)], llm_engine=llm_engine)
```
Comment (pcuenca) [inline agents.md:215]: So the LLM knows it has to generate JSON output because we are using `ReactJsonAgent`? It may be cool to mention it just before or after the snippet, as well as the iterative nature of the process.

File: agents.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+Action:
+{
+  "action": "retriever",
+  "action_input": {"query": "LORA finetuning script"}
+}
+```
+
+### Using a simple multi-agent setup 🤝 for efficient web browsing
+
+We want to build an agent and test it on the GAIA benchmark ([Mialon et al. 2023](https://huggingface.co/papers/2311.12983)). GAIA is an extremely difficult benchmark, with most questions requiring several steps of reasoning using different tools. A specifically difficult requirement is to have a powerful web browser, able to navigate to pages with specific constraints: discovering pages using the website’s inner navigation, selecting specific articles in time...
```
Comment (pcuenca) [inline agents.md:250]:
```suggestion
In this example, we want to build an agent and test it on the GAIA benchmark ([Mialon et al. 2023](https://huggingface.co/papers/2311.12983)). GAIA is an extremely difficult benchmark, with most questions requiring several steps of reasoning using different tools. A specifically difficult requirement is to have a powerful web browser, able to navigate to pages with specific constraints: discovering pages using the website’s inner navigation, selecting specific articles in time...
```

File: agents.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+
+Since this browsing requires diving deeper into subpages, thus scrolling through lots of text tokens that will not necessary for the higher-level task-solving, we prefer to give the web-browsing sub-tasks to a specialized web surfer agent. We provide it with some tools to browse the web and a specific prompt (check the repo to find specific implementations).
```
Comment (pcuenca) [inline agents.md]:
```suggestion
Web browsing requires diving deeper into subpages and scrolling through lots of text tokens that will not be necessary for the higher-level task-solving. We assign the web-browsing sub-tasks to a specialized web surfer agent. We provide it with some tools to browse the web and a specific prompt (check the repo to find specific implementations).
```

File: agents.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    inputs = {
+        "question": {
+            "description": "Your question, as a natural language sentence. You are talking to an agent, so provide them with as much context as possible.",
+            "type": "text",
+        }
+    }
+    output_type = "text"
+
+    def forward(self, question: str) -> str:
+        return surfer_agent.run(question)
```
Comment (pcuenca) [inline agents.md:296]:
```suggestion
        return websurfer_agent.run(question)
```

File: agents.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+```
+
+Then we initialize the task-solving agent this search tool and a specific prompt (basically we added a few modifications over the default prompt for a `ReactCodeAgent` , but you still could use the base prompt, it works well):
```
Comment (pcuenca) [inline agents.md:299]:
```suggestion
Then we initialize the task-solving agent with this search tool that uses a specific prompt (basically we added a few modifications over the default prompt for a `ReactCodeAgent`, but you could still use the base prompt, it works well):
```

I don't fully understand where the specific prompt is coming from, is it the tool, the agent, or both? The code below uses `ReactCodeAgent` so there appears to be no configuration at that point. Maybe we could link to the place where the default is defined?

File: agents.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+## Testing our agents
+
+Let’s take our agent framework for a spin and benchmark different models with it!
+
+All the code for the experiments below can be found [here](https://github.com/aymeric-roucher/agent_reasoning_benchmark).
+
+### Comparing LLM engines
+
+The `agents_reasoning_benchmark` is a small - but mighty- reasoning test for evaluating agent performance. The idea is that the choice the tools you use with your agents can radically alter performance for certain tasks. So this benchmark restricts the set of tools used to a calculator and a basic search tool. And we picked questions from several datasets that could be solved using only these two tools
```
Comment (pcuenca) [inline agents.md:365]:
```suggestion
The `agents_reasoning_benchmark` is a small - but mighty- reasoning test for evaluating agent performance. The idea is that the choice of tools you use with your agents can radically alter performance for certain tasks. So this benchmark restricts the set of tools used to a calculator and a basic search tool. We picked questions from several datasets that could be solved using only these two tools.
```

File: agents.md [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+This benchmark was already used (in an alpha version) and explained in more detail in [our earlier blog post](https://huggingface.co/blog/open-source-llms-as-agents). Here we try 3 different engines: [Mixtral-8x7B](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1), [Llama3-70B](https://huggingface.co/meta-llama/Meta-Llama-3-70B-Instruct), and [GPT-4 Turbo](https://platform.openai.com/docs/models).
+
+<p align="center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/agents/aggregate_score.png" alt="benchmark of agent performances" width=90%>
+</p>
+
+The results are shown above - as the average of two complete runs for more precision. We also tested [Command-R+](https://huggingface.co/CohereForAI/c4ai-command-r-plus) and [Mixtral-8x22B](https://huggingface.co/mistralai/Mixtral-8x22B-Instruct-v0.1), but do not show them for clarity.
+
+**Llama3-70B-Instruct leads the Open-Source models: it is on par with GPT-4**, and it’s especially strong in a `ReactCodeAgent`  thanks to Llama3’s strong coding performance!
```
Comment (pcuenca) [inline agents.md:379]:
```suggestion
**Llama3-70B-Instruct leads the Open-Source models: it is on par with GPT-4**, and it’s especially strong in a `ReactCodeAgent`  thanks to Llama 3’s strong coding performance!
```

File: agents.md [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+
+Comparing JSON- and Code-based agents is interesting: for less powerful models like Mixtral-8x7B, Code-based agents do not perform well, since the LLM often fails to generate good code. But this version catches up at higher levels: in our experience, on the full GAIA test with Llama3-70B-Instruct, the Code version even outperforms the JSON. So that is what we used for the next step: testing on the complete GAIA benchmark.
+
+### Climbing up the GAIA Leaderboard with a multi-modal agent
+
+[GAIA](https://huggingface.co/datasets/gaia-benchmark/GAIA) ([Mialon et al., 2023](https://huggingface.co/papers/2311.12983)) is an extremely difficult benchmark. You can see in the `agent_reasoning_benchmark` above that models still do not perform well although we cherry-picked tasks that they could solve with basic tools. To get food performance we need an agent with specific tools:
```
Comment (pcuenca) [inline agents.md:385]:
```suggestion
[GAIA](https://huggingface.co/datasets/gaia-benchmark/GAIA) ([Mialon et al., 2023](https://huggingface.co/papers/2311.12983)) is an extremely difficult benchmark. You can see in the `agent_reasoning_benchmark` above that models still do not perform well although we cherry-picked tasks that they could solve with basic tools. To get good performance we need an agent with specific tools:
```

File: agents.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+
+- `SearchTool`: the web browser defined above.
+- `TextInspectorTool`: open documents as text files and return their content.
+- `SpeechToTextTool`: transcribe audio files to text. We use the default tool based on [distil-whisper](https://huggingface.co/distil-whisper/distil-large-v3).
+- `VisualQATool`: visualize images. For these we use the shiny new [Idefics2-8b-chatty](https://huggingface.co/HuggingFaceM4/idefics2-8b-chatty)!
```
Comment (pcuenca) [inline agents.md:389]:
```suggestion
- `VisualQATool`: analyze images visually. For these we use the shiny new [Idefics2-8b-chatty](https://huggingface.co/HuggingFaceM4/idefics2-8b-chatty)!
```

File: agents.md [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+    memory_verbose=True,
+)
+```
+
+<p align="center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/agents/leaderboard.png" alt="GAIA leaderboard" width=90%>
+</p>
+
+And after some time needed to complete the 165 questions, we submit our result to the [GAIA Leaderboard](https://huggingface.co/spaces/gaia-benchmark/leaderboard), and… 🥁🥁🥁 
+⇒ Our agent comes at the 4th place: it beats many GPT4-based agents, and is now the reigning contender for the Open-Source category!
```
Comment (pcuenca) [inline agents.md:418]: Maybe mention that it's there in the Validation tab, or additionally put a screenshot here. Is the cherry-picking of questions that was mentioned above compatible with being in the leaderboard? (sorry, I don't know how it works)

File: agents.md [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+
+
+## Conclusion
+
+We will keep improving this package in the coming months. We have already identified several exciting paths:
+
+- agent sharing options
```
Comment (pcuenca) [inline agents.md:424]: Not mentioned in the post, is it? (just in the summary)

**Review by pcuenca — APPROVED** [review]
Amazing, very interesting! 🔥

File: agents.md [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+
+⇒ 💡 We aim for the code to be **simple and clear, and for common attributes like the final prompt and tools to be transparent**.
+
+⇒ 🤝 We add **sharing options** to boost community agents.
+
+⇒ 💪 **Extremely performant new agent framework**, allowing a Llama3-70B agent to outperform GPT-4 based agents in the GAIA Leaderboard!
+
+🚀 Go try it out and climb ever higher on the GAIA leaderboard!
+
+## Table of Contents
```
[the remark this answers is not in the document]
Comment (aymeric-roucher) [reply agents.md:25]: Oh thank you that's useful info! Next time I won't bother with building it then.

File: agents.md [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+authors:
+  - user: m-ric
+  - user: lysandre
+---
+
+# License to Call: Releasing Transformers Agents v2.0
+
+### TL;DR
+
+We are releasing Transformers Agents 2.0!
```
[the remark this answers is not in the document]
Comment (aymeric-roucher) [reply agents.md:13]: Yes the docs are being updated as well! But yes it will be cool to point to the new doc anyway.

File: agents.md (hunk 7 above)
Comment (aymeric-roucher) [reply agents.md]: Thank you! Also changed this in other parts of the post.

File: agents.md (hunk 10 above)
Comment (aymeric-roucher) [reply agents.md:420]: The screenshot is on the picture just above this code snippet! Putting it in-between the two sentences ("And after some time..." and "=> Our agent comes at...") to make it more easy to see!

And thank you for the precision of the cherry-picking: indeed this time we did not cherry pick so our validation set is complete. I'll make this clearer in the post.

File: _blog.yml [hunk 14]
```diff
@@ -3935,3 +3935,20 @@
...
     - LLM
+
+- local: agents
+  title: "License to Call: Releasing Transformers Agents v2.0"
+  thumbnail: /blog/assets/agents/thumbnail.png
+  author: m-ric
+  date: May 07, 2024
+  tags:
+    - nlp
+    - leaderboard
```
[the remark this answers is not in the document]
Comment (aymeric-roucher) [reply _blog.yml]: Thank you, removed!

File: agents.md [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+title: "License to Call: Releasing Transformers Agents v2.0"
+thumbnail: /blog/assets/agents/thumbnail.png
+authors:
+  - user: m-ric
+  - user: lysandre
+---
+
+# License to Call: Releasing Transformers Agents v2.0
+
+### TL;DR
```
[the remark this answers is not in the document]
Comment (aymeric-roucher) [reply agents.md:11]: Should I move the ToC above this section then @pcuenca ?

File: agents.md (hunk 10 above)
Comment (pcuenca) [reply agents.md]: > The screenshot is on the picture

Ah cool, maybe you posted it while I was still reviewing the previous draft, or maybe I just missed it :)

File: agents.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply agents.md:11]: I think it's fine to open with a brief intro, then the ToC guides people on what the rest is about.

File: agents.md (hunk 1 above)
Comment (aymeric-roucher) [reply agents.md:215]: Added a bit more explanation!

File: agents.md (hunk 5 above)
Comment (aymeric-roucher) [reply agents.md:299]: Thank you for pointing this out: I removed this mention of a custom prompt since it was not needed and made things more complicated.

File: agents.md (hunk 11 above)
Comment (aymeric-roucher) [reply agents.md:435]: We already have some agent sharing options mentioned in the summary: in more detail (which is not given in the summary) we can push/load tools from the Hub.
The idea to extend this would be to be able to push/load agents as well! Do you think I should precise that?

Comment (aymeric-roucher) [reply agents.md:435]: (Tagging you @pcuenca for viz)

File: agents.md [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+        )
+```
+
+Now it’s straightforward to create an agent that leverages this tool!
+
+The agent will need these arguments upon initialization:
+- *`tools`*: a list of tools that the agent will be able to call.
+- *`llm_engine`*: the LLM that powers the agent.
+
+Our `llm_engine` must be a callable that takes as input a list of [messages](./chat_templating.) and returns text. It also needs to accept a `stop_sequences` argument that indicates when to stop its generation. For convenience, we directly use the `HfEngine` class provided in the package to get a LLM engine that calls our [Inference API](https://huggingface.co/docs/api-inference/en/index).
```
Comment (pcuenca) [inline agents.md:217]:
```suggestion
Our `llm_engine` must be a callable that takes as input a list of [messages](https://huggingface.co/docs/transformers/main/chat_templating) and returns text. It also needs to accept a `stop_sequences` argument that indicates when to stop its generation. For convenience, we directly use the `HfEngine` class provided in the package to get a LLM engine that calls our [Inference API](https://huggingface.co/docs/api-inference/en/index).
```

File: agents.md (hunk 11 above)
Comment (pcuenca) [reply agents.md:443]: Yes, we can maybe include here what you just said about being able to push/load tools from the Hub and extending to agents in the future :)
